modules/chainheadfix/fix.go

```diff
@@ -0,0 +1,212 @@
+package chainheadfix
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/FactomProject/factomd/common/interfaces"
+	"github.com/FactomProject/factomd/database/databaseOverlay"
+)
+
+// Workers specifies the amounts of worker goroutines used to scan through the database.
+// There is a strong diminishing return for higher values.
+var Workers = 8
+
+// Summary of Chain Head Fixing:
+// To ensure that the database is consistent and that the database entry
+// for "GetChainHead" points to the most recent EBlock of every chain.
+//
+// This is accomplished in a multithreaded three stage process:
+//   1. Iterate over all DBlocks and save the maximum height & keymr of every chain
+//   2. Compare the accurate chain heads of 1. with the database results of chain heads
+//   3. (if enabled) Update the database for any entries that mismatch in 2.
+//
+// The order that DBlocks are iterated over doesn't matter. Currently the limiting factor
+// is the i/o from the database.
+type finder struct {
+	// general
+	db       *databaseOverlay.Overlay
+	log      *log.Entry
+	fixHeads bool
+
+	// keeping track of actual chain heads
+	// results of the first step
+	heads    map[[32]byte]*head
+	headsMtx sync.Mutex
+
+	// multithreaded administrative vars
+	wg     sync.WaitGroup
+	gather chan uint32
+	fix    chan *head
+
+	// counter of how many anomalies were detected
+	errors int64
+
+	// results of the second st ep
+	batchMtx    sync.Mutex
+	batchKeys   []interfaces.IHash
+	batchValues []interfaces.IHash
+}
+
+// retrieve the head struct for a chain id.
+// creates a new head if none is found.
+func (f *finder) get(id interfaces.IHash) *head {
+	f.headsMtx.Lock()
+	var ch *head
+	var ok bool
+	key := id.Fixed()
+	if ch, ok = f.heads[key]; !ok {
+		if ch, ok = f.heads[key]; !ok {
+			ch = newhead(id)
+			f.heads[key] = ch
+		}
+	}
+	f.headsMtx.Unlock()
+	return ch
+}
+
+// step 1
+// launches workers and writes all database heights to worker channel.
+// returns an error if the dblock height could not be determined.
+func (f *finder) gatherHeads(workers int) error {
+	head, err := f.db.FetchDBlockHead()
+	if err != nil {
+		return err
+	}
+
+	// empty database
+	if head == nil {
+		return nil
+	}
+
+	f.wg.Add(workers)
+	for i := 0; i < workers; i++ {
+		go f.gatherer()
```

Comment:
it's a bit bewildering that wg.Add and wg.Wait are here, but wg.Done is buried in the gather method. If I want to check the correctness I need to go back and forth between the 2 pieces of code. A more easy approach for future readers would be for me:
```
go func() {
f.gatherer();
f.wg.Done();
}
```
so that all the WG logic is contained in a single place.

Reply:
I don't want to argue on any coding style though, we don't have that luxury unfortunately, but I wanted to share my thoughts.

Reply:
I like that suggestion a lot, now I don't need to have wg be a struct variable.